[TensorPipe] Add options for agent, including backend killswitches

lw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#40162 [TensorPipe] Add options for agent, including backend killswitches**

The only public option is `num_worker_threads`. The other ones are private (as indicated by the leading underscore, is that enough?) and allow to specify a different set and order of transports/channels. These can thus be used to disable a backend (by not specifying it) or by forcing one (by raising its priority). They can therefore be used to work around defective backends, in case we'll find any post-release.

Differential Revision: [D22090661](https://our.internmc.facebook.com/intern/diff/D22090661/)

lw committed:
[TensorPipe] Add options for agent, including backend killswitches

The only public option is `num_worker_threads`. The other ones are private (as indicated by the leading underscore, is that enough?) and allow to specify a different set and order of transports/channels. These can thus be used to disable a backend (by not specifying it) or by forcing one (by raising its priority). They can therefore be used to work around defective backends, in case we'll find any post-release.

Differential Revision: [D22090661](https://our.internmc.facebook.com/intern/diff/D22090661/)

[ghstack-poisoned]

mrshenli reviewed (commented):
Shall we add some Python tests for this?

lw committed:
Update on "[TensorPipe] Add options for agent, including backend killswitches"


The only public option is `num_worker_threads`. The other ones are private (as indicated by the leading underscore, is that enough?) and allow to specify a different set and order of transports/channels. These can thus be used to disable a backend (by not specifying it) or by forcing one (by raising its priority). They can therefore be used to work around defective backends, in case we'll find any post-release.

Differential Revision: [D22090661](https://our.internmc.facebook.com/intern/diff/D22090661/)

[ghstack-poisoned]

lw commented:
Where would they go? Are there tests for the ProcessGroup agent's options that I can refer to?

mrshenli commented:
> Where would they go?

Looks like it should go `TensorPipeAgentRpcTest` in `rpc/rpc_test.py`?

> Are there tests for the ProcessGroup agent's options that I can refer to?

Yep, you can search for `ProcessGroupRpcBackendOptions` in that file.

osalpekar commented on torch/csrc/distributed/rpc/tensorpipe_agent.h:
```diff
@@ -18,9 +18,61 @@ namespace rpc {
+      for (const std::string& transportName : transports.value()) {
+        TORCH_CHECK(
+            TensorPipeTransportRegistry()->Has(transportName),
+            "Unknwon transport: ",
```
nit: Unknown

osalpekar commented on torch/csrc/distributed/rpc/tensorpipe_agent.h:
```diff
@@ -18,9 +18,61 @@ namespace rpc {
+      for (const std::string& channelName : channels.value()) {
+        TORCH_CHECK(
+            TensorPipeChannelRegistry()->Has(channelName),
+            "Unknwon channel: ",
```
nit: Unknown

lw committed:
Update on "[TensorPipe] Add options for agent, including backend killswitches"


The only public option is `num_worker_threads`. The other ones are private (as indicated by the leading underscore, is that enough?) and allow to specify a different set and order of transports/channels. These can thus be used to disable a backend (by not specifying it) or by forcing one (by raising its priority). They can therefore be used to work around defective backends, in case we'll find any post-release.

Differential Revision: [D22090661](https://our.internmc.facebook.com/intern/diff/D22090661/)

[ghstack-poisoned]

lw commented:
I added some tests, and fixed the typos.

I had to create copies of the tests because at the moment we don't have a decorator that allows to, say, only enable a test on two agents but not on the others. This is hard because we're checking for ProcessGroup and TensorPipe in different ways (because they are "injected" differently). It shouldn't be too hard to address it in the planned refactoring, but probably after v1.6.

mrshenli commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -475,9 +475,22 @@ PyObject* rpc_init(PyObject* /* unused */) {
+          py::arg("_channels") = optional<std::vector<std::string>>(),
           py::arg("rpc_timeout") = kDefaultRpcTimeoutSeconds,
-          py::arg("init_method") = kDefaultInitMethod);
+          py::arg("init_method") = kDefaultInitMethod)
```
We will need a doc string for it. And then add this type to `docs/source/rpc.rst`.

lw replied:
I wanted to do a pass over all the documentation needed for TensorPipe in a subsequent diff.

mrshenli commented on torch/csrc/distributed/rpc/tensorpipe_agent.cpp:
```diff
@@ -217,22 +218,51 @@ void TensorPipeAgent::startImpl() {
+      auto iter =
+          std::find(opts_.channels->begin(), opts_.channels->end(), key);
+      if (iter == opts_.channels->end()) {
+        continue;
```
This means absent transport/channels are not registered (killed)?

lw replied:
Exactly. This is in fact the main purpose of these fields. If you prefer I can do it the other way around, and have these fields be "blacklists" for backends, rather than "whitelists".

mrshenli commented on torch/csrc/distributed/rpc/tensorpipe_agent.h:
```diff
@@ -18,9 +18,61 @@ namespace rpc {
+        "num_worker_threads must be positive, got ",
+        numWorkerThreads);
+
+    if (transports.has_value()) {
+      for (const std::string& transportName : transports.value()) {
+        TORCH_CHECK(
+            TensorPipeTransportRegistry()->Has(transportName),
+            "Unknown transport: ",
+            transportName);
+      }
+    }
+
+    if (channels.has_value()) {
+      for (const std::string& channelName : channels.value()) {
+        TORCH_CHECK(
+            TensorPipeChannelRegistry()->Has(channelName),
+            "Unknown channel: ",
+            channelName);
+      }
+    }
```
Do we need to verify transports and channels match?

lw replied:
There's pros and cons. The pros of verifying is that we would catch a mistyped name and avoid it silently slipping through. The cons are that if we ever remove a backend then all the users configs that specify it will become invalid. These are private options and are intended to be short-lived (until we find a better long-term solution), so personally I feel the former concern outweighs the latter. What do you think?

mrshenli replied:
Sure, let's then keep this as-is.

mrshenli commented on torch/csrc/distributed/rpc/tensorpipe_agent.h:
```diff
@@ -18,9 +18,61 @@ namespace rpc {
+  }
+
+  int numWorkerThreads;
+  optional<std::vector<std::string>> transports;
+  optional<std::vector<std::string>> channels;
```
these can be const?

mrshenli commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -2212,6 +2221,27 @@ def test_set_and_get_num_send_recv_threads(self):
 
+    # FIXME Merge this test and the previous one
+    @dist_init(setup_rpc=False)
+    @_skip_unless_tensorpipe_agent
```
@osalpekar would I be correct if I assume that moving these tests to `class TensorPipeAgentRpcTest` will save use this skip decorator?

osalpekar replied:
Yes, we should move the TP-only tests to `class TensorPipeAgentRpcTest`

lw commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -475,9 +475,22 @@ PyObject* rpc_init(PyObject* /* unused */) {
+          "num_worker_threads", &TensorPipeRpcBackendOptions::numWorkerThreads);
+
+  module.attr("_DEFAULT_NUM_WORKER_THREADS") =
+      py::cast(kDefaultNumWorkerThreads);
```
What do you mean? For a start, they have different values: NUM_SEND_RECV threads is 4 by default, NUM_WORKER_THREADS is 16. Also, they control different aspects. I am not so familiar with the ProcessGroup agent but I think those threads deal both with I/O (receiving and sending) and with the execution of the requests. With TensorPipe, the worker threads just execute the requests, while separate threads deal with I/O. (There's no controls for those threads yet, because they are specific to transports/channels, and we don't have a way to configure options for them yet).

mrshenli replied:
I see

lw commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -475,9 +475,22 @@ PyObject* rpc_init(PyObject* /* unused */) {
+              optional<std::vector<std::string>>,
+              float,
+              std::string>(),
+          py::arg("num_worker_threads") = kDefaultNumWorkerThreads,
```
See answer above. I think they mean different things.

lw committed:
Update on "[TensorPipe] Add options for agent, including backend killswitches"


The only public option is `num_worker_threads`. The other ones are private (as indicated by the leading underscore, is that enough?) and allow to specify a different set and order of transports/channels. These can thus be used to disable a backend (by not specifying it) or by forcing one (by raising its priority). They can therefore be used to work around defective backends, in case we'll find any post-release.

Differential Revision: [D22090661](https://our.internmc.facebook.com/intern/diff/D22090661/)

[ghstack-poisoned]

lw commented:
Made those fields const and moved the TP-only tests to the TensorPipeAgentRpcTest class
